modules/core/src/main/java/io/fluo/webindex/core/models/URL.java

```diff
@@ -0,0 +1,270 @@
+    if (logError) {
+      log.error(msg);
+    } else {
+      log.debug(msg);
+    }
+    throw new IllegalArgumentException(msg);
+  }
+
+  public static URL from(String rawUrl, Function<String, String> domainFromHost,
+      Function<String, Boolean> isValidHost) {
+
+    if (rawUrl.contains(PAGE_ID_SEP)) {
+      badUrl(false, "Skipping raw URL as it contains '" + PAGE_ID_SEP +"':" + rawUrl);
+    }
+
+    String trimUrl = rawUrl.trim();
+    if (trimUrl.length() < 8) {
+      badUrl(false, "Raw URL is too short to start with valid protocol: " + rawUrl);
+    }
+
+    String urlNoProto = "";
+    boolean secure = false;
+    int port = 80;
+    if (trimUrl.substring(0, 7).equalsIgnoreCase(HTTP_PROTO)) {
+      urlNoProto = trimUrl.substring(7);
+    } else if (trimUrl.substring(0, 8).equalsIgnoreCase(HTTPS_PROTO)) {
+      urlNoProto = trimUrl.substring(8);
+      secure = true;
+      port = 443;
+    } else {
+      badUrl(false, "Raw URL does not start with valid protocol: " + rawUrl);
+    }
+
+    String hostPort;
+    String[] args = urlNoProto.split(URL_SEP_REGEX, 2);
+    String path;
+    String sep;
+    if (args.length == 2) {
+      hostPort = args[0].toLowerCase();
+      int sepIndex = args[0].length();
+      sep = urlNoProto.substring(sepIndex, sepIndex + 1);
+      path = sep + args[1];
+    } else {
+      hostPort = urlNoProto.toLowerCase();
+      path = "/";
+    }
+
+    args = hostPort.split(":", 2);
+    String host;
+    if (args.length == 2) {
+      host = args[0];
+      try {
+        port = Integer.parseInt(args[1]);
+      } catch (NumberFormatException ignored) {
+      }
+    } else {
+      host = hostPort;
+    }
+
+    if (host.isEmpty()) {
+      badUrl(false, "Raw URL cannot have empty host: " + rawUrl);
+    }
+
+    String domain = host;
+    if (!isValidIP(host)) {
+      if (!isValidHost.apply(host)) {
+        badUrl(false, "Raw URL (" + rawUrl + ") has invalid host: " + host);
+      }
+      domain = domainFromHost.apply(host);
+    }
+
+    return new URL(domain, host, path, port, secure);
+  }
+
+  public static boolean isValid(String rawUrl, Function<String, String> domainFromHost,
+      Function<String, Boolean> isValidHost) {
+    try {
+      from(rawUrl, domainFromHost, isValidHost);
+      return true;
+    } catch (Exception e) {
+      return false;
+    }
+  }
+
+  public static boolean isValidIP(String host) {
+    return validator.isValid(host);
+  }
+
+  public static String reverseHost(String host) {
+    String[] hostArgs = host.split("\\.");
+    ArrayUtils.reverse(hostArgs);
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < hostArgs.length - 1; i++) {
+      sb.append(hostArgs[i]);
+      sb.append(".");
+    }
+    sb.append(hostArgs[hostArgs.length - 1]);
+    if (host.endsWith(".")) {
+      sb.append(".");
+    }
+    return sb.toString();
+  }
+
+  public boolean hasIPAddr() {
+    return isValidIP(host);
```

Comment:
Do you know if this call to isValidIP is expensive?  If so I was thinking could cache the result since this URL object seems immutable.